### ISSUE TITLE
[no-unstable-nested-components] Fat arrow functions nested in component are not detected by lint rule. 

### ISSUE THREAD

**Issue by An631** [jsx-eslint/eslint-plugin-react#3227]
Regarding: no-unstable-nested-components.md rule

Fat arrow functions that are nested in a component and hence part of its re-render path, have the same state reinitialization issue. Unfortunately, the lint rule plugin is not catching them.

For example (taken from the [documentation example](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unstable-nested-components.md)), the fat arrow function for **InternalSubComponent** below will cause the same re-render issue in its internal state so it should be caught by the linter:

```ts
function FaultyReusableComponent(props) {

    const InternalSubComponent = (props) => (
        <div>
            <BlockText>InternalSubComponent</BlockText>
            {props.children}
        </div>
    );

    return (
        <div className="reusable-component reusable-component--faulty">
            <h2>Faulty Reusable Component</h2>
            <InternalSubComponent children={props.children}/>
        </div>
    );
}
```

**Comment by AriPerkkio** [comment]
Could you share a complete repro for this issue - especially your `.eslintrc`?

This case is covered by unit testing:

https://github.com/yannickcr/eslint-plugin-react/blob/d8602acf4d17e317b34290fe7c95ae0212f4030f/tests/lib/rules/no-unstable-nested-components.js#L638-L652

I also quickly tested this in minimal repro and it is reported as error:

<details>

<summary>Minimal repro</summary>

```json
{
  "root": true,
  "parserOptions": {
    "ecmaVersion": 2020,
    "sourceType": "module",
    "ecmaFeatures": { "jsx": true }
  },
  "plugins": ["react"],
  "rules": {
    "react/no-unstable-nested-components": "error"
  }
}
```

```json
{
  "scripts": {
    "lint": "eslint test.js"
  },
  "devDependencies": {
    "eslint": "^8.10.0",
    "eslint-plugin-react": "^7.29.2",
    "react": "^17.0.2"
  }
}
```

```js
// test.js
function FaultyReusableComponent(props) {
  const InternalSubComponent = (props) => (
    <div>
      <BlockText>InternalSubComponent</BlockText>
      {props.children}
    </div>
  );

  return (
    <div className="reusable-component reusable-component--faulty">
      <h2>Faulty Reusable Component</h2>
      <InternalSubComponent children={props.children} />
    </div>
  );
}
```

```
 ➜ /workspaces/eslint-demo-repro (main ✗) $ yarn lint
yarn run v1.22.10
warning package.json: No license field
$ eslint test.js
Warning: React version not specified in eslint-plugin-react settings. See https://github.com/yannickcr/eslint-plugin-react#configuration .

/workspaces/eslint-demo-repro/test.js
  2:32  error  Declare this component outside parent component "FaultyReusableComponent" or memoize it  react/no-unstable-nested-components

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

</details>

**Comment by jeiea** [comment]
Just one question. I thought the followings should be accepted or denied together, but it did not.

Is it expected? If then what is the reason?

```
function Other() {
  return null;
}

function Foo() {
  const now = useRef(new Date());
  const whyThisAcceptable = (props) => (
    <ChallengeDayTile now={now.current} {...props} />
  );

  return (
    <div>
      <Other s={whyThisAcceptable} /> // no error
      <Other s={(props) => <ChallengeDayTile now={now.current} {...props} />} /> // react/no-unstable-nested-components
    </div>
  );
}
```

**Comment by AriPerkkio** [comment]
> ```jsx
>   const whyThisAcceptable = (props) => (
>     <Some now={now.current} {...props} />
>   );
> ```

This is not a React component. If you try to create React element from this, by using it as `<whyThisAcceptable />` or `React.createElement(whyThisAcceptable, {}, null)` React will throw error. All components must be uppercase. This rule will simply ignore all lowercase functions which return JSX.

> ```jsx
> <Other s={(props) => <Some now={now.current} {...props} />} /> // react/no-unstable-nested-components
> ```

This rule has some assumptions. It will prevent you from passing inline components as props by default - unless you prefix the prop's name with `render`. This means that `header={() => <div />}` is forbidden, while `renderHeader={() => <div />}` is allowed. You can disable all component-in-prop-checks by using the [`allowAsProps` rule option](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-unstable-nested-components.md#rule-options).

**Comment by jeiea** [comment]
I tested `React.createElement(whyThisAcceptable, {}, null)` and it seems not throwing error. I still have doubt about this rule but it's not important. Thanks for the comment, I misunderstood where render prefix can be used :)

**Comment by ljharb** [comment]
This seems answered. If you're still having problems, please file a new issue with repro code and we'll take a look at it :-)
